File: shells/ios-template/WefterBridge/WefterBridge.swift
import Foundation

/// Framework-owned registry of live plugin instances, keyed by plugin name. Lets a native
/// component that isn't the dispatcher's own view controller, a presented view controller, a
/// URL handler — reach back into a plugin to call `emit()`, without every plugin author
/// hand-rolling their own static accessor.
public enum WefterBridge {
    private static var plugins: [String: WefterPlugin] = [:]
    private static let lock = NSLock()

    static func register(_ pluginId: String, plugin: WefterPlugin) {
        lock.lock()
        defer { lock.unlock() }
        plugins[pluginId] = plugin
    }

    public static func plugin(withId pluginId: String) -> WefterPlugin? {
        lock.lock()
        defer { lock.unlock() }
        return plugins[pluginId]
    }
}

File: shells/android-template/app/src/main/java/dev/wefter/bridge/WefterBridge.kt
package dev.wefter.bridge

import java.util.concurrent.ConcurrentHashMap

/**
 * Framework-owned registry of live plugin instances, keyed by plugin name. Lets a native component
 * that isn't the dispatcher's own Activity, a launched Activity, a Service, a BroadcastReceiver —
 * reach back into a plugin to call `emit()`/dispatch a hook, without every plugin author
 * hand-rolling their own static accessor.
 */
object WefterBridge {
    private val plugins = ConcurrentHashMap<String, WefterPlugin>()

    internal fun register(pluginId: String, plugin: WefterPlugin) {
        plugins[pluginId] = plugin
    }

    fun getPlugin(pluginId: String): WefterPlugin? = plugins[pluginId]
}
